lib/encode.js

```diff
@@ -13,40 +20,88 @@ import {inspect} from './util.js';
+ * @param {Map} options.stringTable - A map of string values and
+ *  their associated CBOR-LD integer values.
+ * @param {Map} options.urlSchemeTable - A map of URL schemes to
+ *  their associated CBOR-LD integer values.
+ * @param {Map} options.typedLiteralTable - A map of JSON-LD types
+ *  to maps that map values of that type to their associated CBOR-LD
+ *  integer values.
+ * @param {diagnosticFunction} options.diagnose - A function that, if
+ *  provided, is called with diagnostic information.
+ * @param {Map} options.appContextMap - For use with the legacy value of
+ *  varintTagValue.
+ * @param {number} options.compressionMode - For use with the legacy value of
+ *  varintValue.
  * @returns {Promise<Uint8Array>} - The encoded CBOR-LD bytes.
  */
 export async function encode({
-  jsonldDocument, documentLoader, appContextMap = new Map(),
-  compressionMode = 1, diagnose
+  jsonldDocument, documentLoader, varintValue = -1,
+  keywordsTable = new Map(),
+  stringTable = new Map(),
+  urlSchemeTable = new Map(),
+  typedLiteralTable = new Map(),
+  diagnose,
+  appContextMap = new Map(),
+  compressionMode = 1
 } = {}) {
-  if(!(compressionMode === 0 || compressionMode === 1)) {
-    throw new TypeError(
-      '"compressionMode" must be "0" (no compression) or "1" ' +
-      'for compression mode version 1.');
-  }
-
-  // 0xd9 == 11011001
-  // 110 = CBOR major type 6
-  // 11001 = 25, 16-bit tag size (65536 possible values)
-  // 0x05 = always the first 8-bits of a CBOR-LD tag
-  // compressionMode = last 8-bits of a CBOR-LD tag indicating compression type
-  const prefix = new Uint8Array([0xd9, 0x05, compressionMode]);
+  let prefix;
   let suffix;
+  // process using legacy values if appContextMap passed
+  if(varintValue == -1) {
+    prefix = new Uint8Array([0xd9, 0x05, compressionMode]);
 
-  if(compressionMode === 0) {
-    // handle uncompressed CBOR-LD
-    suffix = cborg.encode(jsonldDocument);
+    const stringTableWithAppContext = new Map(STRING_TABLE);
+
+    for(const [key, value] of appContextMap) {
+      stringTableWithAppContext.set(key, value);
+    }
+    if(compressionMode === 0) {
+      // handle uncompressed CBOR-LD
+      suffix = cborg.encode(jsonldDocument);
+    } else {
+      // compress CBOR-LD
+      const compressor = new Compressor({
+        documentLoader,
+        keywordsTable: new Map(KEYWORDS_TABLE),
+        stringTable: stringTableWithAppContext,
+        urlSchemeTable: new Map(URL_SCHEME_TABLE),
+        typedLiteralTable: new Map(TYPED_LITERAL_TABLE)
+      });
+      suffix = await compressor.compress({jsonldDocument, diagnose});
+    }
   } else {
-    // compress CBOR-LD
-    const compressor = new Compressor({documentLoader, appContextMap});
-    suffix = await compressor.compress({jsonldDocument, diagnose});
+    // otherwise, process with new structure/values
+    if(!(varintValue >= 1)) {
+      throw new TypeError(
+        '"varintValue" must be an integer greater than 0');
+    }
```

Comment:
We might end up moving this check higher/elsewhere, but this seemed easier to match against the string message:

```suggestion
    if(!(varintValue > 0)) {
      throw new TypeError(
        '"varintValue" must be an integer greater than 0.');
    }
```